### PR TITLE
add WarmupStepLR

### PR DESCRIPTION
**Pull request by HiHippie**
该pr为LiBai添加WarmupStepLR支持，以及修复test_warmup_exponential测试代码error

- [x] WarmupStepLR代码
- [x] unittest代码
- [x] 测试(在vit上做了一些iter的训练，WarmupStepLR学习率衰减正常)
- [x] 修复test_warmup_exponential测试代码error

`Auto-scaling the config to train.train_iter=100, train.warmup_iter=2 step_size=10`

```
[06/27 08:45:41 lb.engine.trainer]: Starting training from iteration 0
[06/27 08:45:43 lb.utils.events]:  iteration: 0/100  consumed_samples: 16  total_loss: 2.344  data_time: 0.2230 s/iter  lr: 0.00e+00  
[06/27 08:45:43 lb.utils.events]:  eta: 0:00:42  iteration: 1/100  consumed_samples: 32  total_loss: 2.373  data_time: 0.2223 s/iter  lr: 5.00e-04  
[06/27 08:45:44 lb.utils.events]:  eta: 0:00:26  iteration: 2/100  consumed_samples: 48  total_loss: 2.401  time: 0.2740 s/iter  data_time: 0.2175 s/iter total_throughput: 58.38 samples/s lr: 1.00e-03  
[06/27 08:45:44 lb.utils.events]:  eta: 0:00:26  iteration: 3/100  consumed_samples: 64  total_loss: 2.706  time: 0.2739 s/iter  data_time: 0.2169 s/iter total_throughput: 58.41 samples/s lr: 1.00e-03  
[06/27 08:45:44 lb.utils.events]:  eta: 0:00:26  iteration: 4/100  consumed_samples: 80  total_loss: 2.693  time: 0.2773 s/iter  data_time: 0.2180 s/iter total_throughput: 57.70 samples/s lr: 1.00e-03  
[06/27 08:45:44 lb.utils.events]:  eta: 0:00:26  iteration: 5/100  consumed_samples: 96  total_loss: 2.648  time: 0.2784 s/iter  data_time: 0.2162 s/iter total_throughput: 57.47 samples/s lr: 1.00e-03  
[06/27 08:45:45 lb.utils.events]:  eta: 0:00:26  iteration: 6/100  consumed_samples: 112  total_loss: 2.602  time: 0.2788 s/iter  data_time: 0.2170 s/iter total_throughput: 57.39 samples/s lr: 1.00e-03  
[06/27 08:45:45 lb.utils.events]:  eta: 0:00:25  iteration: 7/100  consumed_samples: 128  total_loss: 2.57  time: 0.2788 s/iter  data_time: 0.2174 s/iter total_throughput: 57.38 samples/s lr: 1.00e-03  
[06/27 08:45:45 lb.utils.events]:  eta: 0:00:25  iteration: 8/100  consumed_samples: 144  total_loss: 2.538  time: 0.2749 s/iter  data_time: 0.2142 s/iter total_throughput: 58.21 samples/s lr: 1.00e-03  
[06/27 08:45:46 lb.evaluation.evaluator]: with eval_iter 10, reset total samples 10000 to 2560
[06/27 08:45:46 lb.evaluation.evaluator]: Start inference on 2560 samples
[06/27 08:46:02 lb.evaluation.evaluator]: Total valid samples: 2560
[06/27 08:46:02 lb.evaluation.evaluator]: Total inference time: 0:00:07.512320 (0.002940 s / iter per device, on 4 devices)
[06/27 08:46:02 lb.evaluation.evaluator]: Total inference pure compute time: 0:00:01 (0.000398 s / iter per device, on 4 devices)
[06/27 08:46:02 lb.engine.default]: Evaluation results for CIFAR10Dataset in csv format:
[06/27 08:46:02 lb.evaluation.utils]: copypaste: Acc@1=14.6875
[06/27 08:46:02 lb.evaluation.utils]: copypaste: Acc@5=49.84375
[06/27 08:46:02 lb.engine.hooks]: Saved first model at 14.68750 @ 9 steps
[06/27 08:46:03 lb.utils.checkpoint]: Saving checkpoint to output_unittest/test_vit/model_best
[06/27 08:46:09 lb.utils.events]:  eta: 0:00:25  iteration: 9/100  consumed_samples: 160  total_loss: 2.536  time: 0.2766 s/iter  data_time: 0.2146 s/iter total_throughput: 57.84 samples/s lr: 1.00e-03  
[06/27 08:46:09 lb.utils.events]:  eta: 0:00:24  iteration: 10/100  consumed_samples: 176  total_loss: 2.534  time: 0.2692 s/iter  data_time: 0.2082 s/iter total_throughput: 59.43 samples/s lr: 1.00e-04  
[06/27 08:46:10 lb.utils.events]:  eta: 0:00:24  iteration: 11/100  consumed_samples: 192  total_loss: 2.467  time: 0.2686 s/iter  data_time: 0.2083 s/iter total_throughput: 59.57 samples/s lr: 1.00e-04  
[06/27 08:46:10 lb.utils.events]:  eta: 0:00:24  iteration: 12/100  consumed_samples: 208  total_loss: 2.428  time: 0.2693 s/iter  data_time: 0.2085 s/iter total_throughput: 59.42 samples/s lr: 1.00e-04  
[06/27 08:46:10 lb.utils.events]:  eta: 0:00:23  iteration: 13/100  consumed_samples: 224  total_loss: 2.414  time: 0.2665 s/iter  data_time: 0.2071 s/iter total_throughput: 60.03 samples/s lr: 1.00e-04  
[06/27 08:46:10 lb.utils.events]:  eta: 0:00:23  iteration: 14/100  consumed_samples: 240  total_loss: 2.428  time: 0.2654 s/iter  data_time: 0.2069 s/iter total_throughput: 60.28 samples/s lr: 1.00e-04  
[06/27 08:46:11 lb.utils.events]:  eta: 0:00:23  iteration: 15/100  consumed_samples: 256  total_loss: 2.414  time: 0.2641 s/iter  data_time: 0.2061 s/iter total_throughput: 60.59 samples/s lr: 1.00e-04  
[06/27 08:46:11 lb.utils.events]:  eta: 0:00:22  iteration: 16/100  consumed_samples: 272  total_loss: 2.401  time: 0.2621 s/iter  data_time: 0.2048 s/iter total_throughput: 61.05 samples/s lr: 1.00e-04  
[06/27 08:46:11 lb.utils.events]:  eta: 0:00:22  iteration: 17/100  consumed_samples: 288  total_loss: 2.389  time: 0.2597 s/iter  data_time: 0.2030 s/iter total_throughput: 61.62 samples/s lr: 1.00e-04  
[06/27 08:46:11 lb.utils.events]:  eta: 0:00:21  iteration: 18/100  consumed_samples: 304  total_loss: 2.378  time: 0.2581 s/iter  data_time: 0.2020 s/iter total_throughput: 62.00 samples/s lr: 1.00e-04  
[06/27 08:46:12 lb.evaluation.evaluator]: with eval_iter 10, reset total samples 10000 to 2560
[06/27 08:46:12 lb.evaluation.evaluator]: Start inference on 2560 samples
[06/27 08:46:28 lb.evaluation.evaluator]: Total valid samples: 2560
[06/27 08:46:28 lb.evaluation.evaluator]: Total inference time: 0:00:07.433129 (0.002909 s / iter per device, on 4 devices)
[06/27 08:46:28 lb.evaluation.evaluator]: Total inference pure compute time: 0:00:00 (0.000322 s / iter per device, on 4 devices)
[06/27 08:46:28 lb.engine.default]: Evaluation results for CIFAR10Dataset in csv format:
[06/27 08:46:28 lb.evaluation.utils]: copypaste: Acc@1=14.4921875
[06/27 08:46:28 lb.evaluation.utils]: copypaste: Acc@5=48.515625
[06/27 08:46:28 lb.engine.hooks]: Not saving as latest eval score for Acc@1 is 14.49219, not better than best score 14.68750 @ iteration 9.
[06/27 08:46:28 lb.utils.events]:  eta: 0:00:20  iteration: 19/100  consumed_samples: 320  total_loss: 2.37  time: 0.2574 s/iter  data_time: 0.2016 s/iter total_throughput: 62.16 samples/s lr: 1.00e-04  
[06/27 08:46:29 lb.utils.events]:  eta: 0:00:19  iteration: 20/100  consumed_samples: 336  total_loss: 2.362  time: 0.2561 s/iter  data_time: 0.1996 s/iter total_throughput: 62.48 samples/s lr: 1.00e-05  
[06/27 08:46:29 lb.utils.events]:  eta: 0:00:19  iteration: 21/100  consumed_samples: 352  total_loss: 2.37  time: 0.2556 s/iter  data_time: 0.1984 s/iter total_throughput: 62.59 samples/s lr: 1.00e-05  
[06/27 08:46:29 lb.utils.events]:  eta: 0:00:19  iteration: 22/100  consumed_samples: 368  total_loss: 2.362  time: 0.2548 s/iter  data_time: 0.1975 s/iter total_throughput: 62.79 samples/s lr: 1.00e-05  
[06/27 08:46:29 lb.utils.events]:  eta: 0:00:18  iteration: 23/100  consumed_samples: 384  total_loss: 2.362  time: 0.2542 s/iter  data_time: 0.1964 s/iter total_throughput: 62.95 samples/s lr: 1.00e-05  
[06/27 08:46:30 lb.utils.events]:  eta: 0:00:18  iteration: 24/100  consumed_samples: 400  total_loss: 2.362  time: 0.2542 s/iter  data_time: 0.1955 s/iter total_throughput: 62.95 samples/s lr: 1.00e-05  
[06/27 08:46:30 lb.utils.events]:  eta: 0:00:18  iteration: 25/100  consumed_samples: 416  total_loss: 2.364  time: 0.2546 s/iter  data_time: 0.1956 s/iter total_throughput: 62.84 samples/s lr: 1.00e-05  
[06/27 08:46:30 lb.utils.events]:  eta: 0:00:18  iteration: 26/100  consumed_samples: 432  total_loss: 2.362  time: 0.2542 s/iter  data_time: 0.1944 s/iter total_throughput: 62.93 samples/s lr: 1.00e-05  
[06/27 08:46:30 lb.utils.events]:  eta: 0:00:17  iteration: 27/100  consumed_samples: 448  total_loss: 2.362  time: 0.2538 s/iter  data_time: 0.1930 s/iter total_throughput: 63.04 samples/s lr: 1.00e-05  
[06/27 08:46:31 lb.utils.events]:  eta: 0:00:17  iteration: 28/100  consumed_samples: 464  total_loss: 2.362  time: 0.2537 s/iter  data_time: 0.1935 s/iter total_throughput: 63.08 samples/s lr: 1.00e-05 
```

### CODE REVIEW

File: libai/scheduler/lr_scheduler.py [hunk 1]
```diff
@@ -98,6 +98,47 @@ def WarmupCosineAnnealingLR(
...
+        train_epoch (int): Total training epochs.
+        warmup_factor (float): The warmup factor.
+        warmup_iter (int): The number of warmup steps.
+        step_size (int): Period of learning rate decay.
+        gamma (float, optional): Multiplicative factor of learning rate decay. Defaults to 0.1.
+        warmup_method (str, optional): The method of warmup, you can choose "linear" or "constant".
+            In linear mode, the multiplication factor starts with warmup_factor in the first
+            epoch and then inreases linearly to reach 1. Defaults to "linear".
+    """
+    step_size = (step_size / train_epoch) * max_iter
```
Comment (HiHippie) [inline libai/scheduler/lr_scheduler.py:126]: StepLR文档中step_size是按照epoch来计数的，但我看LiBai是每个iter执行一次scheduler.step()。

所以这一行是把epoch_step_size转化为iter_step_size，我理解这样会更方便理解去设置StepLR参数(与文档保持一致)，然后直接根据epoch设置step_size也方便些。

不过已有的MultiStepLR的milestone也是按照iter去设置的。

所以这里问下有没有必要加这一行处理~
@CPFLAME @Ldpe2G

Comment (CPFLAME) [reply libai/scheduler/lr_scheduler.py:128]: libai会自动修正 用户的入参 `train_epoch`, 转换为`train_iter`
https://github.com/Oneflow-Inc/libai/blob/de2c68f2692760e5de87ebb815541a98d1b8ebe7/libai/engine/default.py#L682-L685

而且在default_trainer里面  是先进行了cfg中`train_iter`的变换, 再`build_scheduler`.

https://github.com/Oneflow-Inc/libai/blob/de2c68f2692760e5de87ebb815541a98d1b8ebe7/libai/engine/default.py#L298-L303

所以我倾向于这里直接写`train_iter`相关的参数就好


但我不是很确定这里的在`build_scheduler`这里 train_iter的改变是否生效了.  需要放到libai里面跑一下即可

Comment (CPFLAME) [reply libai/scheduler/lr_scheduler.py:128]: 跑一下验证的方式应该也比较简单, 只设置'train_epoch' 然后train_iter设置成0. 
然后观察lr的变化规律就可以了.

Comment (HiHippie) [reply libai/scheduler/lr_scheduler.py:128]: 好的

train_iter在`build_lr_scheduler`中作为`max_iter`生效了的。

如果不做iter->epoch处理的话，我想在config中指定step_size=200epoch 是不是只能手动算一下对应多少个iter这样子~？

Comment (CPFLAME) [reply libai/scheduler/lr_scheduler.py]: > 如果不做iter->epoch处理的话，我想在config中指定step_size=200epoch 是不是只能手动算一下对应多少个iter这样子~？
对 应该是.

如果你想在config里面也实现 step_size=200epoch 的话, 建议在config里面额外加一个参数  `epoch_step_size` 之类的, 然后在`default_trainer.py`里面加上把它转换成 `iter_step_size`的代码,  然后真正送给你的lr_scheduler的是你的`iter_step_size`

Comment (HiHippie) [reply libai/scheduler/lr_scheduler.py]: > > 如果不做iter->epoch处理的话，我想在config中指定step_size=200epoch 是不是只能手动算一下对应多少个iter这样子~？
> > 对 应该是.
> 
> 如果你想在config里面也实现 step_size=200epoch 的话, 建议在config里面额外加一个参数 `epoch_step_size` 之类的, 然后在`default_trainer.py`里面加上把它转换成 `iter_step_size`的代码, 然后真正送给你的lr_scheduler的是你的`iter_step_size`

嗯嗯~目前就是这么实现的。在default_trainer.py的build_lr_scheduler中做了转换。

File: libai/scheduler/lr_scheduler.py [hunk 2]
```diff
@@ -98,6 +98,44 @@ def WarmupCosineAnnealingLR(
     return warmup_cosine_annealing_lr
 
 
+def WarmupStepLR(
+    optimizer: flow.optim.Optimizer,
+    max_iter: int,
```
Comment (Ldpe2G) [inline libai/scheduler/lr_scheduler.py:103]: 这个变量没用到？

Comment (HiHippie) [reply libai/scheduler/lr_scheduler.py:103]: > 这个变量没用到？

是的，我参考WarmupMultiStepLR等其他函数来实现的，都是保留max_iter等参数(可能是有没用到的)，然后加上额外需要的参数。

Comment (Ldpe2G) [reply libai/scheduler/lr_scheduler.py:103]: 是为了和torch的实现对齐么？如果不是是不是可以都删了

Comment (HiHippie) [reply libai/scheduler/lr_scheduler.py:103]: 我来看下删掉会不会有其他影响

Comment (HiHippie) [reply libai/scheduler/lr_scheduler.py:103]: https://github.com/Oneflow-Inc/libai/blob/de2c68f2692760e5de87ebb815541a98d1b8ebe7/libai/engine/default.py#L667-L720

`auto_scale_hyperparams` 720行这里调用了max_iter，所以应该是不能删掉的。

Comment (Ldpe2G) [reply libai/scheduler/lr_scheduler.py:103]: 好的

File: tests/test_scheduler.py [hunk 3]
```diff
@@ -105,18 +133,18 @@ def test_warmup_exponential(self):
...
-            for idx in range(max_iters - warmup_iters):
+            for idx in range(warmup_iters, max_iters+1):
                 valid_values.append(base_lr * (gamma ** idx))
             return valid_values
 
         for _ in range(30):
             sched.step()
             lrs.append(opt.param_groups[0]["lr"])
-        self.assertTrue(np.allclose(lrs[:5], [0.005, 1.004, 2.003, 3.002, 4.001]))
+        self.assertTrue(np.allclose(lrs[:5], [0.005, 0.00401, 0.0030199999999999997, 0.00203, 0.0010399999999999997]))
```
Comment (HiHippie) [inline tests/test_scheduler.py:143]: 这里不通过是因为这里的warmup中lr是下降的

Comment (Ldpe2G) [reply tests/test_scheduler.py:143]: warmup 时候 lr 是下降的？这个是正常行为么？

File: tests/test_scheduler.py [hunk 4]
```diff
@@ -105,18 +133,18 @@ def test_warmup_exponential(self):
 
         def _get_exponential_lr(base_lr, gamma, max_iters, warmup_iters):
             valid_values = []
-            for idx in range(max_iters - warmup_iters):
+            for idx in range(warmup_iters, max_iters+1):
```
[the remark this answers is not in the document]
Comment (Ldpe2G) [reply tests/test_scheduler.py:136]: 感觉应该是对的，这里是收集从warmup_iters到max_iter之间的lr

File: tests/test_scheduler.py (hunk 3 above)
Comment (HiHippie) [reply tests/test_scheduler.py:143]: image

image

应该是正常行为，end_factor < start_factor时候就是下降的


https://github.com/Oneflow-Inc/oneflow/blob/b17a9cd6b930b5817c63623fb682bd708377a93b/python/oneflow/nn/optimizer/warmup_lr.py#L149-L158

这段代码调用test_warmup_exponential更新的lr计算end_factor，导致其变小。

Comment (Ldpe2G) [reply tests/test_scheduler.py:143]: 好的

**Review by Ldpe2G — APPROVED** [review]

**Review by CPFLAME — APPROVED** [review]

File: libai/scheduler/lr_scheduler.py [hunk 5]
```diff
@@ -203,7 +239,7 @@ def WarmupPolynomialLR(
             epoch and then inreases linearly to reach 1. Defaults to "linear".
     """
     polynomial_lr = flow.optim.lr_scheduler.PolynomialLR(
-        optimizer, steps=max_iter, end_learning_rate=end_learning_rate, power=power, cycle=cycle
+        optimizer, decay_batch=max_iter, end_learning_rate=end_learning_rate, power=power, cycle=cycle
```
Comment (HiHippie) [inline libai/scheduler/lr_scheduler.py:242]: 因为我的oneflow版本不是最新的，所以本地测试通过，但准备merge的时候发现这里的key_arg有变动。
参考 https://github.com/Oneflow-Inc/models/pull/349 这个pr

所以这里同步做了修改
